monkey/infection_monkey/exploit/tools/brute_force_exploiter.py

```diff
@@ -0,0 +1,209 @@
+import logging
+from contextlib import suppress
+from io import BytesIO
+from pathlib import PurePath
+from time import time
+from typing import Callable, Iterable, MutableSet, Set
+
+from common import OperatingSystem
+from common.agent_events import ExploitationEvent, PropagationEvent
+from common.credentials import Credentials
+from common.event_queue import IAgentEventPublisher
+from common.types import AgentID, Event
+from infection_monkey.exploit import IAgentBinaryRepository
+from infection_monkey.i_puppet import ExploiterResultData, TargetHost
+from infection_monkey.utils.threading import interruptible_iter
+
+from . import (
+    IRemoteAccessClient,
+    IRemoteAccessClientFactory,
+    RemoteAuthenticationError,
+    RemoteCommandExecutionError,
+    RemoteFileCopyError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class BruteForceExploiter:
+    """
+    An exploiter that brute-forces credentials and propagates the Monkey agent
+
+    Operates on any exploit client that implements `IRemoteAccessClient`. An
+    instance of `IRemoteAccessClientFactory` must be provided to create the
+    exploit client.
+    """
+
+    def __init__(
+        self,
+        exploiter_name: str,
+        agent_id: AgentID,
+        destination_path: PurePath,
+        build_command: Callable[[OperatingSystem, PurePath, PurePath], str],
+        exploit_client_factory: IRemoteAccessClientFactory,
+        get_credentials: Callable[[], Iterable[Credentials]],
+        agent_binary_repository: IAgentBinaryRepository,
+        agent_event_publisher: IAgentEventPublisher,
+        tags: Set[str],
```

Comment:
```suggestion
        event_tags: Set[str],
```

Reply:
There should be a unit test that verifies all events are published with these tags